### PR TITLE
Merge model compression dev branch to master

### PR DESCRIPTION
**Pull request by liuzhe-lz**

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -10,6 +10,11 @@ jobs:
   steps:
   - script: python3 -m pip install --upgrade pip setuptools --user
     displayName: 'Install python tools'
+  - script: |
```
Comment (xuehui1991) [inline azure-pipelines.yml]: why we need to merge the master to dev-mc?

Comment (liuzhe-lz) [reply azure-pipelines.yml]: This is from dev to master

File: docs/en_US/Compressor/Overview.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+```
+
+Pytorch code
+```
+config = [{'sparsity':0.8,'support_type': 'default'}]
+pruner = nni.compressors.torch_compressor.LevelPruner(config)
+pruner(model)
+```
+
+You can use other compression algorithms in the package of `nni.compressors`. The algorithms are implemented in both Pytorch and Tensorflow, under `nni.compressors.torch_compressor` and `nni.compressors.tf_compressor` respectively. You can refer to [Pruner](./Pruner) and [Quantizer](./Quantizer) for detail description of supported algorithms.
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Overview.md]: It seems the package torch_compressor has been renamed in code.

Comment (liuzhe-lz) [reply docs/en_US/Compressor/Overview.md]: Related PR has just been merged.

File: docs/en_US/Compressor/Overview.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
+# Compressor
+NNI provides an easy-to-use toolkit to help user design and use compression algorithms. It supports Tensorflow and PyTorch with unified interface. For users to compress their models, they only need to add several lines in their code. There are some popular model compression algorithms built-in in NNI. Users could further use NNI's auto tuning power to find the best compressed model, which is detailed in [Auto Model Compression](./AutoCompression.md). On the other hand, users could easily customize their new compression algorithms using NNI's interface, refer to the tutorial [here](#customize-new-compression-algorithms).
+
+## Supported algorithms
+We have provided two naive compression algorithms and four popular ones for users, including three pruning algorithms and three quantization algorithms:
+
+|Name|Brief Introduction of Algorithm|
+|---|---|
+| [Level Pruner](./Pruner.md#level-pruner) | Pruning the specified ratio on each weight based on absolute values of weights |
+| [AGP Pruner](./Pruner.md#agp-pruner) | To prune, or not to prune: exploring the efficacy of pruning for model compression. [Reference Paper](https://arxiv.org/abs/1710.01878)|
```
Comment (xuehui1991) [inline docs/en_US/Compressor/Overview.md]: can you explain the AGP pruner in short sentence?

File: docs/en_US/Compressor/Overview.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    {
+        'sparsity': 0.8,
+        'op_types': 'default'
+    },
+    {
+        'sparsity': 0.6,
+        'op_names': ['op_name1', 'op_name2']
+    },
+    {
+        'exclude': True,
```
Comment (xuehui1991) [inline docs/en_US/Compressor/Overview.md]: why we need this field?

File: docs/en_US/Compressor/Overview.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    
+    def bind_model(self, model):
+        #preprocess model
+```
+
+__[TODO]__ Will add another member function `quantize_layer_output`, as some quantization algorithms also quantize layers' output.
+
+### Usage of user customized compression algorithm
+
+__[TODO]__ ...
```
Comment (xuehui1991) [inline docs/en_US/Compressor/Overview.md]: And how？

File: docs/en_US/Compressor/Overview.md (hunk 4 above)
Comment (liuzhe-lz) [reply docs/en_US/Compressor/Overview.md]: This means "do not compress op_name3"

File: docs/en_US/Compressor/Overview.md (hunk 5 above)
Comment (liuzhe-lz) [reply docs/en_US/Compressor/Overview.md]: In fact we have not considered customized compressor yet...

File: docs/en_US/Compressor/Overview.md (hunk 3 above)
Comment (liuzhe-lz) [reply docs/en_US/Compressor/Overview.md]: I dunno how the algorithms works... Maybe @tanglang96 can help?

Comment (tanglang96) [reply docs/en_US/Compressor/Overview.md]: AGP means Automated Gradual Pruning, prune with gradually increasing sparsity while training (given initial sparsity and final sparsity) until the final sparsity reached,  the smallest magnitude weights are masked with zeros.

File: docs/en_US/Compressor/Overview.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+Some algorithms generate mask based on training progress, i.e., epoch number. We provide `update_epoch` for the pruner to be aware of the training progress.
+
+Some algorithms may want global information for generating masks, for example, all weights of the model (for statistic information), model optimizer's information. NNI supports this requirement using `bind_model`. `bind_model` receives the complete model, thus, it could record any information (e.g., reference to weights) it cares about. Then `step` can process or update the information according to the algorithm. You can refer to [source code of built-in algorithms](https://github.com/microsoft/nni/tree/master/src/sdk/pynni/nni/compressors) for example implementations.
+
+### Quantization algorithm
+
+The interface for customizing quantization algorithm is similar to that of pruning algorithms. The only difference is that `calc_mask` is replaced with `quantize_weight`. `quantize_weight` directly returns the quantized weights rather than mask, because for quantization the quantized weights cannot be obtained by applying mask.
+
+```
```
Comment (squirrelsc) [inline docs/en_US/Compressor/Overview.md]:
```python

File: docs/en_US/Compressor/Overview.md (hunk 3 above)
Comment (QuanluZhang) [reply docs/en_US/Compressor/Overview.md]: @liuzhe-lz better to add this sentence

File: docs/en_US/Compressor/Overview.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        'op_names': ['op_name3']
+    }
+]
+```
+It means following the algorithm's default setting for compressed operations with sparsity 0.8, but for `op_name1` and `op_name2` use sparsity 0.6, and please do not compress `op_name3`.
+
+### Other APIs
+
+Some compression algorithms use epochs to control the progress of compression, and some algorithms need to do something after every minibatch. Therefore, we provide another two APIs for users to invoke. One is `update_epoch`, you can use it as follows:
+
```
Comment (xuehui1991) [inline docs/en_US/Compressor/Overview.md]: could you explain when the user needs to use 'update_epoch'?

File: docs/en_US/Compressor/Pruner.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+## AGP Pruner
+In [To prune, or not to prune: exploring the efficacy of pruning for model compression](https://arxiv.org/abs/1710.01878), authors Michael Zhu and Suyog Gupta provide an algorithm to prune the weight gradually.
+
+>We introduce a new automated gradual pruning algorithm in which the sparsity is increased from an initial sparsity value si (usually 0) to a final sparsity value sf over a span of n pruning steps, starting at training step t0 and with pruning frequency ∆t:
+![](../../img/agp_pruner.png)
+>The binary weight masks are updated every ∆t steps as the network is trained to gradually increase the sparsity of the network while allowing the network training steps to recover from any pruning-induced loss in accuracy. In our experience, varying the pruning frequency ∆t between 100 and 1000 training steps had a negligible impact on the final model quality. Once the model achieves the target sparsity sf , the weight masks are no longer updated. The intuition behind this sparsity function in equation
+
+### Usage
+You can prune all weight from %0 to 80% sparsity in 10 epoch with the code below.
```
Comment (xuehui1991) [inline docs/en_US/Compressor/Pruner.md]: 0%?

File: examples/model_compress/main_torch_quantizer.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    # like thaht quantizer.compress(model)
+    
+
+    optimizer = torch.optim.SGD(model.parameters(), lr = 0.01, momentum = 0.5)
+    for epoch in range(10):
+        print('# Epoch {} #'.format(epoch))
+        train(model, device, train_loader, optimizer)
+        test(model, device, test_loader)
+
+
```
Comment (xuehui1991) [inline examples/model_compress/main_torch_quantizer.py]: remove the empty line.

File: src/sdk/pynni/nni/compression/tensorflow/builtin_pruners.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    weights to achieve a preset level of network sparsity.
+
+    Michael Zhu and Suyog Gupta, "To prune, or not to prune: exploring the
+    efficacy of pruning for model compression", 2017 NIPS Workshop on Machine
+    Learning of Phones and other Consumer Devices,
+    https://arxiv.org/pdf/1710.01878.pdf
+    """
+    def __init__(self, config_list):
+        """
+            Configure Args
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/compression/tensorflow/builtin_pruners.py]: could you follow the docstring format in other tuner...

File: docs/en_US/Compressor/Overview.md (hunk 3 above)
Comment (liuzhe-lz) [reply docs/en_US/Compressor/Overview.md]: Seems too long. I added "automated gradual pruning".

File: docs/en_US/Compressor/Overview.md (hunk 7 above)
Comment (liuzhe-lz) [reply docs/en_US/Compressor/Overview.md]: Added an example compressor using `update_epoch`. But failed to find an example using `step`.

File: docs/en_US/Compressor/Pruner.md (hunk 8 above)
Comment (liuzhe-lz) [reply docs/en_US/Compressor/Pruner.md]: Yes, seems that's how the algorithm works.

File: src/sdk/pynni/nni/compression/tensorflow/builtin_pruners.py (hunk 10 above)
Comment (liuzhe-lz) [reply src/sdk/pynni/nni/compression/tensorflow/builtin_pruners.py]: I found compressors' docstring format is different from other modules (like tuner).
But I don't know what format we are using exactly.
@SparkSnail Any idea?

**Comment by liuzhe-lz** [comment]
Updates are going to PR #1590

**Comment by liuzhe-lz** [comment]
Due to failures in our pipeline, the patch PR #1590 is not able to merge in time.
I'll merge this PR first and redirect #1590 to master.
